File: client-sdk/typescript/src/agents.ts
// `Agents` — owner of the heartbeat wildcard subscription and the entry
// point for discovery. Orchestrates subscribe-before-discover per §8.5
// and returns live `Agent` handles.
//
// Construct with a pre-opened `NatsConnection`:
//
//   import { connect } from "@nats-io/transport-node";
//   import { Agents } from "@synadia-ai/agents";
//
//   const nc = await connect({ servers: "nats://localhost:4222" });
//   const agents = new Agents({ nc });
//   const found = await agents.discover({ timeoutMs: 2_000 });
//   for await (const msg of await found[0]!.prompt("hi")) { ... }
//
// The caller owns `nc`. `agents.close()` tears down SDK-owned state only;
// closing the underlying connection is the caller's responsibility.
//
// Sender identity (extension): pass `identity: { signer, name }` to sign
// every `prompt` / `status` request; pass `identity: {}` explicitly for an
// unsigned claim. Omitting `identity` performs no automatic lookup and sends
// no header. `selfId()` is the connection's own agent ID; `signSender` /
// `publishSigned` / `requestSigned` sign arbitrary publishes (JetStream
// included); `resolveSender(id)` is the reverse lookup (agent ID → the
// agent that registered it, `id_sig` verified, TTL-cached).
//
// Prompt interceptors: `interceptors: [...]` runs each one before every
// prompt an `Agent` from this client publishes (see `prompt/interceptor.ts`).

import type { Msg, NatsConnection } from "@nats-io/nats-core";
import type { Agent } from "./agent.js";
import type { AgentInfo } from "./discovery/agent-info.js";
import {
  discoverAgents,
  lookupAgentInstance,
  pingInstance,
  type DiscoverOptions,
} from "./discovery/srv-ping.js";
import { HeartbeatTracker, type Liveness } from "./heartbeat/tracker.js";
import { type HeartbeatPayload } from "./heartbeat/payload.js";
import type { AgentId } from "./identity/agent-id.js";
import { IdentityContext, type IdentityOptions } from "./identity/context.js";
import { SenderResolver } from "./identity/resolve-sender.js";
import { refreshSelfId, selfId } from "./identity/self-id.js";
import { serializeSenderHeader } from "./identity/sender-header.js";
import {
  NATS_MSG_ID_HEADER,
  signedPublishHeaders,
  signedSenderHeader,
  toBytes,
  type SignedPublishOptions,
} from "./identity/signed-publish.js";
import { type Logger, SILENT_LOGGER } from "./internal/logger.js";
import type { PromptInterceptor } from "./prompt/interceptor.js";

/** Default per-stream inactivity timeout (§6.6) — 60 seconds. */
export const DEFAULT_STREAM_INACTIVITY_TIMEOUT_MS = 60_000;

/** Default `requestSigned()` timeout — 2 seconds. */
export const DEFAULT_REQUEST_SIGNED_TIMEOUT_MS = 2_000;

export { NATS_MSG_ID_HEADER, type SignedPublishOptions };

export interface AgentsOptions {
  /** A pre-connected `NatsConnection`. Caller retains ownership. */
  readonly nc: NatsConnection;
  /** Default per-stream inactivity timeout in milliseconds. Default: 60_000. */
  readonly streamInactivityTimeoutMs?: number;
  /** Pluggable logger. Default: silent. */
  readonly logger?: Logger;
  /**
   * Sender-identity configuration. Omit for no lookup/header; pass `{}`
   * explicitly for an unsigned claim, or provide a signer.
   */
  readonly identity?: IdentityOptions;
  /**
   * TTL of the `$SRV.INFO.agents` index behind {@link Agents.resolveSender},
   * in milliseconds; `0` enumerates on every call. Default 10 000.
   */
  readonly resolveTtlMs?: number;
  /**
   * Prompt interceptors, run in order before every prompt an `Agent` from
   * this client publishes — each may publish signed messages of its own and
   * add envelope fields and headers. Default: none, and prompts are exactly
   * what the protocol sends.
   */
  readonly interceptors?: ReadonlyArray<PromptInterceptor>;
}

export interface SignedRequestOptions extends SignedPublishOptions {
  /** Request timeout in milliseconds. Default {@link DEFAULT_REQUEST_SIGNED_TIMEOUT_MS}. */
  readonly timeoutMs?: number;
}

export class Agents {
  readonly #nc: NatsConnection;
  readonly #tracker: HeartbeatTracker;
  readonly #logger: Logger;
  readonly #streamInactivityTimeoutMs: number;
  readonly #identity: IdentityContext | undefined;
  readonly #interceptors: ReadonlyArray<PromptInterceptor>;
  readonly #resolver: SenderResolver;
  readonly #closeController = new AbortController();
  #closed = false;

  constructor(options: AgentsOptions) {
    this.#nc = options.nc;
    this.#logger = options.logger ?? SILENT_LOGGER;
    this.#streamInactivityTimeoutMs =
      options.streamInactivityTimeoutMs ?? DEFAULT_STREAM_INACTIVITY_TIMEOUT_MS;
    // Omission is meaningful: no automatic lookup and no sender header.
    // An explicit empty object opts into an unsigned claim.
    this.#identity =
      options.identity !== undefined
        ? new IdentityContext(options.nc, options.identity)
        : undefined;
    // Copied: a caller mutating its array afterwards changes nothing here.
    this.#interceptors = Object.freeze([...(options.interceptors ?? [])]);
    // Validates `resolveTtlMs` up front (throws `IdentityError`).
    this.#resolver = new SenderResolver(
      options.nc,
      options.resolveTtlMs !== undefined ? { ttlMs: options.resolveTtlMs } : {},
    );
    this.#tracker = new HeartbeatTracker(options.nc, this.#logger);
  }

  /** The underlying NATS connection. */
  get connection(): NatsConnection {
    return this.#nc;
  }

  /** Default per-stream inactivity timeout applied to every `Agent.prompt()`. */
  get streamInactivityTimeoutMs(): number {
    return this.#streamInactivityTimeoutMs;
  }

  /** The prompt interceptors every `Agent` from this client runs, in order. */
  get interceptors(): ReadonlyArray<PromptInterceptor> {
    return this.#interceptors;
  }

  /**
   * Abort signal that fires when `close()` is called. Callers that construct
   * `Agent` instances outside of `discover()` (e.g. materialising one from a
   * heartbeat + `$SRV.INFO.agents.<id>` lookup) SHOULD pass this to the
   * `Agent` constructor so in-flight streams on those handles are aborted
   * when the `Agents` client is torn down, matching what `discover()` does.
   */
  get closeSignal(): AbortSignal {
    return this.#closeController.signal;
  }

  /**
   * Discover protocol-compliant agents reachable on the NATS connection.
   * Returns a live `Agent[]` — each entry is directly callable via `.prompt()`.
   *
   * The first call to `discover()` lazily starts the heartbeat wildcard
   * subscription BEFORE publishing `$SRV.PING`, enforcing §8.5 automatically.
   * With an explicit identity option that can send a signed or unsigned
   * claim, it also *starts* the connection's identity lookup
   * (fire-and-forget) so the first `prompt()` usually finds it memoised.
   * With identity omitted, or with unsigned claims explicitly disabled and
   * no signer, discovery performs no identity work.
   *
   * Two instances of the same logical agent (same `(agent, owner, name)`)
   * show up as separate entries with distinct `instanceId`s; callers who
   * want one-per-logical-agent can group with `Map.groupBy`.
   */
  async discover(opts: DiscoverOptions = {}): Promise<Agent[]> {
    this.#ensureOpen();
    this.#identity?.kickoff();
    if (!this.#tracker.isStarted) {
      // tracker.start() flushes internally so the SUB is at the server before
      // we send $SRV.PING (§8.5 subscribe-before-discover).
      await this.#tracker.start();
    }
    return discoverAgents(
      this.#nc,
      this.#streamInactivityTimeoutMs,
      this.#closeController.signal,
      opts,
      this.#identity,
      this.#interceptors,
      this.#logger,
    );
  }

  /**
   * Ensure the heartbeat wildcard subscription is established. Normally
   * called implicitly by `discover()` / `onHeartbeat()`; use this when you
   * want to start tracking before either.
   */
  async startTracking(): Promise<void> {
    this.#ensureOpen();
    await this.#tracker.start();
  }

  /** Current passively-tracked liveness for an instance, or `null` if no heartbeat seen. */
  liveness(instanceId: string): Liveness | null {
    return this.#tracker.liveness(instanceId);
  }

  /**
   * Subscribe to heartbeats for a single instance. Returns an unsubscribe
   * function. Tracker is started lazily if needed — call `startTracking()`
   * beforehand (and await it) if you need to guarantee the subscription is
   * live before a specific moment.
   */
  onHeartbeat(instanceId: string, listener: (payload: HeartbeatPayload) => void): () => void {
    if (!this.#tracker.isStarted) {
      // Fire-and-forget: lazy start. Callers who need determinism use
      // startTracking() first. (Flush happens inside tracker.start.)
      void this.#tracker.start();
    }
    return this.#tracker.onHeartbeat(instanceId, listener);
  }

  /**
   * On-demand reachability check for a single instance via
   * `$SRV.PING.agents.{instanceId}` (§8.4).
   *
   * Returns `true` as soon as any response arrives within the timeout;
   * `false` on timeout.
   */
  async ping(instanceId: string, opts: { timeoutMs?: number } = {}): Promise<boolean> {
    this.#ensureOpen();
    return pingInstance(this.#nc, instanceId, opts);
  }

  /**
   * Targeted `$SRV.INFO.agents.<instanceId>` lookup. Returns a constructed
   * {@link Agent} for an already-known instance id, or `null` if the
   * instance doesn't reply within the timeout / replies with malformed
   * metadata. The returned `Agent` shares this client's connection and
   * close signal, so closing the `Agents` client cancels in-flight streams
   * on the looked-up handle too.
   *
   * Use this when you already have an `instance_id` (e.g. from a heartbeat
   * payload) and want to materialise an `Agent` without re-running a
   * full discovery scan.
   */
  async lookupInstance(
    instanceId: string,
    opts: { timeoutMs?: number } = {},
  ): Promise<Agent | null> {
    this.#ensureOpen();
    return lookupAgentInstance(
      this.#nc,
      instanceId,
      this.#streamInactivityTimeoutMs,
      this.#closeController.signal,
      opts,
      this.#identity,
      this.#interceptors,
      this.#logger,
    );
  }

  // ---------------------------------------------------------------------
  // Sender identity
  // ---------------------------------------------------------------------

  /**
   * The connection's own agent ID (`{account}.{user}`), learned from
   * `$SYS.REQ.USER.INFO` and cached per connection and identity source. A
   * credentials JWT is compared with the live user/account, never used as a
   * substitute for binding. Rejects with {@link NoIdentityError}
   * (no NKEY user — the message names the fix), {@link IdentityUnavailableError}
   * (no answer / permission violation), or {@link IdentityMismatchError}
   * (a configured signer represents a different user/account). Failures are retried
   * after 30 s; `refreshSelfId()` retries at once.
   */
  selfId(): Promise<AgentId> {
    this.#ensureOpen();
    return this.#identity?.selfId() ?? selfId(this.#nc);
  }

  /** Force a fresh identity lookup, discarding the memoised answer. */
  refreshSelfId(): Promise<AgentId> {
    this.#ensureOpen();
    return this.#identity?.refresh() ?? refreshSelfId(this.#nc);
  }

  /**
   * Build a complete `Agent-Sender` header *value* for a publish of
   * `payload` to `subject` — the SDK supplies id, `ts` and a fresh nonce
   * (or `opts.nonce`, see {@link SignedPublishOptions.nonce}). Pass the
   * exact subject and payload you will publish; set `opts.sub` only behind
   * a rename by your own account. Works for any publish, JetStream
   * included (`headers.set("Agent-Sender", value)`).
   *
   * Rejects with {@link SenderSignatureRequiredError} when no signer is
   * configured, with `IdentityError` for a malformed `opts.nonce`, else
   * with the `selfId()` error when the identity is unavailable.
   */
  async signSender(
    subject: string,
    payload: Uint8Array | string,
    opts: Pick<SignedPublishOptions, "sub" | "nonce"> = {},
  ): Promise<string> {
    this.#ensureOpen();
    return serializeSenderHeader(
      await signedSenderHeader(this.#identity, subject, toBytes(payload), opts),
    );
  }

  /**
   * Sign and publish in one step (core `nc.publish`). Sets
   * `Nats-Msg-Id` to the nonce so a JetStream stream's de-duplication
   * window helps consumers. Same error rules as {@link signSender}.
   */
  async publishSigned(
    subject: string,
    payload: Uint8Array | string,
    opts: SignedPublishOptions = {},
  ): Promise<void> {
    this.#ensureOpen();
    const bytes = toBytes(payload);
    const hdrs = await signedPublishHeaders(this.#identity, subject, bytes, opts);
    this.#nc.publish(subject, bytes, { headers: hdrs });
  }

  /**
   * Sign and send a **single-reply** request (`nc.request`) — for
   * services that answer once. Prompt streams go through
   * `Agent.prompt()`. Same error rules as {@link signSender}.
   */
  async requestSigned(
    subject: string,
    payload: Uint8Array | string,
    opts: SignedRequestOptions = {},
  ): Promise<Msg> {
    this.#ensureOpen();
    const bytes = toBytes(payload);
    const hdrs = await signedPublishHeaders(this.#identity, subject, bytes, opts);
    return this.#nc.request(subject, bytes, {
      timeout: opts.timeoutMs ?? DEFAULT_REQUEST_SIGNED_TIMEOUT_MS,
      headers: hdrs,
    });
  }

  /**
   * Reverse lookup (spec "Reverse lookup: from agent ID to agent"): the
   * `AgentInfo` of the instance that registered `id` with a verifying
   * `id_sig`, or `undefined` when no verified instance claims the key —
   * the sender is then not a reachable agent (a human user, a plain
   * service, an agent that is offline). Enumerates `$SRV.INFO.agents`
   * and caches the index for `resolveTtlMs` (default 10 s); discovery is
   * account-local. Identifies, never authorizes.
   */
  resolveSender(id: AgentId | string): Promise<AgentInfo | undefined> {
    this.#ensureOpen();
    return this.#resolver.resolve(id);
  }

  /**
   * Close the SDK-owned state. Cancels all in-flight prompt streams (they
   * throw AbortError from the iterator) and unsubscribes the heartbeat
   * wildcard. The underlying `NatsConnection` is untouched — the caller
   * who opened it is responsible for closing it.
   */
  async close(): Promise<void> {
    if (this.#closed) return;
    this.#closed = true;
    this.#closeController.abort(new Error("@synadia-ai/agents: Agents is closed"));
    await this.#tracker.stop();
  }

  /** True if `close()` has been called. */
  get isClosed(): boolean {
    return this.#closed;
  }

  #ensureOpen(): void {
    if (this.#closed) {
      throw new Error("@synadia-ai/agents: Agents is closed");
    }
  }
}

File: client-sdk/python/src/synadia_ai/agents/agents.py
""":class:`Agents` — caller-side entry point.

Owner of the heartbeat wildcard subscription and the entry point for
discovery. Construct with a pre-opened :class:`~nats.aio.client.Client`::

    import nats
    from synadia_ai.agents import Agents

    nc = await nats.connect(servers="nats://127.0.0.1:4222")
    agents = Agents(nc=nc)
    found = await agents.discover()
    [agent] = found
    async for msg in agent.prompt("hi"):
        ...
    await agents.close()    # SDK state only — does NOT close `nc`
    await nc.close()        # caller owns this

Mirrors the TS SDK's ``Agents`` class (PR #7) field-for-field. The
caller owns ``nc``; :meth:`Agents.close` tears down SDK-owned state
only (heartbeat wildcard sub, in-flight stream cancellation), and the
underlying :class:`~nats.aio.client.Client` is the caller's
responsibility.

Sender identity is off by default: no lookup and no header. Pass
``identity=Identity(signer=…, name=…)`` to sign every ``prompt`` /
``status`` request; an explicit identity without a signer sends an unsigned
claim when the connection has an NKEY identity. :meth:`Agents.self_id` is the connection's own agent
ID; :meth:`Agents.sign_sender` / :meth:`Agents.publish_signed` /
:meth:`Agents.request_signed` sign arbitrary publishes (JetStream
included); :meth:`Agents.resolve_sender` is the reverse lookup.
"""

from __future__ import annotations

import asyncio
from collections.abc import Callable, Mapping
from typing import TYPE_CHECKING

from ._logging import get_logger
from ._request import request_one
from .agent import DEFAULT_PROMPT_MAX_WAIT_S, DEFAULT_STREAM_INACTIVITY_TIMEOUT_S, Agent
from .discovery import (
    DEFAULT_DISCOVER_MAX_WAIT_S,
    DEFAULT_DISCOVER_STALL_S,
    AgentInfo,
    DiscoverFilter,
    discover_agent_infos,
    matches_filter,
    ping_instance,
)
from .errors import SenderSignatureRequiredError
from .heartbeat import HeartbeatListener, HeartbeatTracker, Liveness
from .identity.agent_id import AgentId
from .identity.options import (
    Identity,
    plan_sender_header,
    refresh_self_id_for,
    self_id_for,
)
from .identity.resolve_sender import DEFAULT_RESOLVE_TTL_S, SenderResolver
from .identity.sender_header import (
    AGENT_SENDER_HEADER,
    AgentSenderHeader,
    serialize_sender_header,
)
from .trace import TraceOptions

if TYPE_CHECKING:
    import logging

    from nats.aio.client import Client as NATSClient
    from nats.aio.msg import Msg

log = get_logger(__name__)

# Default `request_signed()` timeout — 2 seconds (mirrors the TS SDK's
# DEFAULT_REQUEST_SIGNED_TIMEOUT_MS).
DEFAULT_REQUEST_SIGNED_TIMEOUT_S: float = 2.0

#: JetStream de-duplication header; `publish_signed` sets it to the nonce.
NATS_MSG_ID_HEADER = "Nats-Msg-Id"


class Agents:
    """Caller-side entry point. One per application; wraps a NATS connection.

    The caller retains ownership of ``nc`` — :meth:`close` does NOT close
    the underlying connection. Pass a custom logger via ``logger=`` to
    surface SDK-internal events through your app's logging stack;
    defaults to ``logging.getLogger("synadia_ai.agents.agents")``.

    ``identity`` configures the sender-identity extension (signer,
    display name, unsigned-claim policy — see
    :class:`~synadia_ai.agents.identity.Identity`; its ``name`` is
    validated up front). ``resolve_ttl_s`` is the TTL of the
    ``$SRV.INFO`` index behind :meth:`resolve_sender` (default 10 s).
    """

    def __init__(
        self,
        *,
        nc: NATSClient,
        stream_inactivity_timeout: float = DEFAULT_STREAM_INACTIVITY_TIMEOUT_S,
        prompt_max_wait_s: float = DEFAULT_PROMPT_MAX_WAIT_S,
        logger: logging.Logger | None = None,
        identity: Identity | None = None,
        resolve_ttl_s: float = DEFAULT_RESOLVE_TTL_S,
        trace: TraceOptions | None = None,
    ) -> None:
        if prompt_max_wait_s <= 0:
            raise ValueError(f"prompt_max_wait_s must be > 0 (got {prompt_max_wait_s!r}).")
        self._nc = nc
        self._stream_inactivity_timeout = stream_inactivity_timeout
        self._prompt_max_wait_s = prompt_max_wait_s
        self._logger = logger if logger is not None else log
        self._identity = identity
        # Omission is meaningful: no trace options, no tracing.
        self._trace = trace
        self._resolver = SenderResolver(nc, ttl_s=resolve_ttl_s)
        self._tracker = HeartbeatTracker(nc)
        # Set when close() is called; passed to every Agent so in-flight
        # prompt streams can short-circuit instead of waiting on a torn-
        # down broker. The shared mux reply-inbox lives on the
        # connection itself (per-nc singleton in `_mux.py`); this Agents
        # does not own it, mirroring the TS SDK's `nc.requestMany`
        # design where the connection holds the mux.
        self._close_event = asyncio.Event()
        self._closed = False
        self._lazy_start_task: asyncio.Task[None] | None = None

    @property
    def connection(self) -> NATSClient:
        """The underlying NATS connection (caller-owned)."""
        return self._nc

    @property
    def stream_inactivity_timeout(self) -> float:
        """Default per-stream inactivity timeout applied to every :meth:`Agent.prompt`."""
        return self._stream_inactivity_timeout

    @property
    def prompt_max_wait_s(self) -> float:
        """Default absolute ceiling for :meth:`Agent.prompt` (overridable per-call)."""
        return self._prompt_max_wait_s

    @property
    def close_event(self) -> asyncio.Event:
        """Event that fires when :meth:`close` is called.

        Pass to :class:`Agent` constructors built outside of :meth:`discover`
        so in-flight streams on those handles abort when this :class:`Agents`
        is torn down — matching what :meth:`discover` does for handles it
        produces.
        """
        return self._close_event

    @property
    def is_closed(self) -> bool:
        """True if :meth:`close` has been called."""
        return self._closed

    @property
    def identity(self) -> Identity | None:
        """The sender-identity options this client was constructed with."""
        return self._identity

    @property
    def trace(self) -> TraceOptions | None:
        """The tracing options this client was constructed with; ``None`` = tracing off."""
        return self._trace

    async def discover(
        self,
        *,
        timeout: float | None = None,
        stall: float = DEFAULT_DISCOVER_STALL_S,
        max_wait: float = DEFAULT_DISCOVER_MAX_WAIT_S,
        filter: DiscoverFilter | None = None,
    ) -> list[Agent]:
        """Discover protocol-compliant agents reachable on the NATS connection.

        Returns a live ``list[Agent]`` — each entry is directly callable
        via :meth:`Agent.prompt`.

        Two strategies:

        - When ``timeout`` is ``None`` (default), the **stall** strategy
          is used: returns ``stall`` seconds after the most recent reply,
          or after ``max_wait`` seconds absolute, whichever fires first.
          Defaults: ``stall=0.2``, ``max_wait=2.0``. Snappy on lightly-
          loaded systems.
        - When ``timeout`` is set, the **timer** strategy is used: waits
          exactly ``timeout`` seconds and returns every responder seen
          in that window. Use for deterministic scans / health checks.

        ``filter`` AND-matches the discovered records by identity
        (``agent``, ``owner``, ``session_name``, ``protocol_version``)
        before the live :class:`Agent` instances are constructed.

        The first call to :meth:`discover` lazily starts the heartbeat
        wildcard subscription BEFORE publishing the discovery PING,
        enforcing §8.5 automatically. Discovery itself never starts sender-
        identity lookup; identity is resolved only by an explicitly enabled
        identity-bearing request.
        """
        self._ensure_open()
        if not self._tracker.is_started:
            await self._tracker.start()
        infos = await discover_agent_infos(
            self._nc,
            timeout_s=timeout,
            stall_s=stall,
            max_wait_s=max_wait,
        )
        return [
            Agent(
                self._nc,
                info,
                stream_inactivity_timeout=self._stream_inactivity_timeout,
                prompt_max_wait_s=self._prompt_max_wait_s,
                close_event=self._close_event,
                identity=self._identity,
                trace=self._trace,
            )
            for info in infos
            if matches_filter(info, filter)
        ]

    async def start_tracking(self) -> None:
        """Ensure the heartbeat wildcard subscription is established.

        Normally called implicitly by :meth:`discover` / :meth:`on_heartbeat`;
        use this when you want to start tracking before either.
        """
        self._ensure_open()
        await self._tracker.start()

    def liveness(self, instance_id: str) -> Liveness | None:
        """Return the passively-tracked liveness for an instance.

        ``None`` until at least one heartbeat has been observed for
        ``instance_id``. The :class:`Liveness` snapshot includes
        ``is_online`` precomputed at read time.
        """
        return self._tracker.liveness(instance_id)

    def on_heartbeat(
        self,
        instance_id: str,
        listener: HeartbeatListener,
    ) -> Callable[[], None]:
        """Subscribe to heartbeats for a single instance.

        Returns an unsubscribe function. The tracker is started lazily if
        needed — call :meth:`start_tracking` first (and await it) when
        you need to guarantee the subscription is live before a specific
        moment.
        """
        if not self._tracker.is_started and self._lazy_start_task is None:
            # Fire-and-forget: lazy start. Callers who need determinism
            # use start_tracking() first. Hold a reference so the task
            # isn't GC'd before it runs.
            self._lazy_start_task = asyncio.create_task(
                self._tracker.start(), name="heartbeat-tracker-lazy-start"
            )
        return self._tracker.on_heartbeat(instance_id, listener)

    async def ping(self, instance_id: str, *, timeout: float = 2.0) -> bool:
        """On-demand reachability check for a single instance (§8.4).

        Sends ``$SRV.PING.agents.{instance_id}`` and returns ``True`` as
        soon as any reply arrives within ``timeout`` seconds; ``False``
        on timeout or when the broker reports no responders.
        """
        self._ensure_open()
        return await ping_instance(self._nc, instance_id, timeout=timeout)

    # --- sender identity -------------------------------------------------

    async def self_id(self) -> AgentId:
        """The connection's own agent ID (``{account}.{user}``).

        Obtained from ``$SYS.REQ.USER.INFO``. A configured signer is bound
        to that live user and account without using the signer-less memo.
        Raises :class:`NoIdentityError` (no NKEY
        user — the message names the fix),
        :class:`IdentityUnavailableError` (no answer / permission
        violation) or :class:`IdentityMismatchError` (a configured signer
        holds a different identity). Signer-less diagnostic lookups and
        failures are memoised; :meth:`refresh_self_id` retries at once.
        """
        self._ensure_open()
        return await self_id_for(self._identity, self._nc)

    async def refresh_self_id(self) -> AgentId:
        """Force a fresh identity lookup, discarding the memoised answer."""
        self._ensure_open()
        return await refresh_self_id_for(self._identity, self._nc)

    async def sign_sender(
        self, subject: str, payload: bytes | str, *, sub: str | None = None
    ) -> str:
        """Build a complete ``Agent-Sender`` header *value* for a publish to ``subject``.

        The SDK supplies the id, ``ts`` and a fresh nonce. Pass the exact
        subject and payload you will publish; set ``sub`` only behind a
        rename by your own account (sign the exporter's subject). Works
        for any publish, JetStream included
        (``headers={"Agent-Sender": value}``).

        Raises :class:`SenderSignatureRequiredError` when no signer is
        configured, else the :meth:`self_id` error when the identity is
        unavailable.
        """
        header = await self._signed_header(subject, _to_bytes(payload), sub)
        return serialize_sender_header(header)

    async def publish_signed(
        self,
        subject: str,
        payload: bytes | str,
        *,
        sub: str | None = None,
        headers: Mapping[str, str] | None = None,
    ) -> None:
        """Sign and publish in one step (core ``nc.publish``).

        Sets ``Nats-Msg-Id`` to the nonce so a JetStream stream's
        de-duplication window helps consumers. ``headers`` are merged in
        (``Agent-Sender`` / ``Nats-Msg-Id`` win). Same error rules as
        :meth:`sign_sender`.
        """
        data = _to_bytes(payload)
        hdrs = await self._signed_headers(subject, data, sub, headers)
        await self._nc.publish(subject, data, headers=hdrs)

    async def request_signed(
        self,
        subject: str,
        payload: bytes | str,
        *,
        sub: str | None = None,
        headers: Mapping[str, str] | None = None,
        timeout_s: float = DEFAULT_REQUEST_SIGNED_TIMEOUT_S,
    ) -> Msg:
        """Sign and send a **single-reply** request on the SDK inbox.

        For services that answer once; prompt streams go through
        :meth:`Agent.prompt`. Same error rules as :meth:`sign_sender`,
        plus :class:`TimeoutError` / :class:`~nats.errors.NoRespondersError`
        from the transport.
        """
        data = _to_bytes(payload)
        hdrs = await self._signed_headers(subject, data, sub, headers)
        return await request_one(self._nc, subject, data, timeout_s=timeout_s, headers=hdrs)

    async def resolve_sender(self, id: AgentId | str) -> AgentInfo | None:
        """Reverse lookup: the verified agent registered under ``id``, or ``None``.

        Enumerates ``$SRV.INFO.agents`` (cached for ``resolve_ttl_s``),
        keeps only instances whose ``id_sig`` verifies over their own
        prompt subject, and returns the matching :class:`AgentInfo`. The
        lookup identifies; it never authorizes.
        """
        self._ensure_open()
        return await self._resolver.resolve(id)

    async def _signed_header(
        self, subject: str, payload: bytes, sub: str | None
    ) -> AgentSenderHeader:
        self._ensure_open()
        if self._identity is None or self._identity.signer is None:
            raise SenderSignatureRequiredError(subject)
        plan = await plan_sender_header(
            self._identity, self._nc, sub if sub is not None else subject, require_signed=True
        )
        if plan is None:  # unreachable with a signer; keeps the type checker honest
            raise SenderSignatureRequiredError(subject)
        return await plan.build(payload)

    async def _signed_headers(
        self, subject: str, payload: bytes, sub: str | None, extra: Mapping[str, str] | None
    ) -> dict[str, str]:
        header = await self._signed_header(subject, payload, sub)
        hdrs: dict[str, str] = dict(extra) if extra else {}
        hdrs[AGENT_SENDER_HEADER] = serialize_sender_header(header)
        if header.nonce is not None:
            hdrs[NATS_MSG_ID_HEADER] = header.nonce
        return hdrs

    async def close(self) -> None:
        """Tear down SDK-owned state. Idempotent.

        Sets :attr:`close_event`, which every in-flight
        :meth:`Agent.prompt` iterator races against — they unblock
        within an event-loop tick instead of waiting on the inactivity
        timeout. Then unsubscribes the heartbeat wildcard. The shared
        reply-inbox mux is **not** torn down here — it lives on the
        connection (per-nc singleton in ``_mux.py``) and dies when the
        caller closes ``nc``. The underlying NATS connection itself is
        also NOT touched — the caller who opened it is responsible
        for closing it.
        """
        if self._closed:
            return
        self._closed = True
        self._close_event.set()
        await self._tracker.stop()

    def _ensure_open(self) -> None:
        if self._closed:
            raise RuntimeError("synadia_ai.agents.Agents is closed")


def _to_bytes(payload: bytes | str) -> bytes:
    return payload.encode("utf-8") if isinstance(payload, str) else payload


__all__ = ["DEFAULT_REQUEST_SIGNED_TIMEOUT_S", "NATS_MSG_ID_HEADER", "Agents"]
